website/server/models/group.js

```diff
@@ -333,6 +333,95 @@ schema.statics.getGroups = async function getGroups (options = {}) {
+
+  switch (info.type) {
+    case 'quest_start':
+      msg = `\`${shared.i18n.t('chatQuestStarted', {questName: questScrolls[info.quest].text(lang)}, lang)}\``;
+      break;
+
+    case 'boss_damage':
+      msg = `\`${shared.i18n.t('chatBossDamage', {username: info.user, bossName: questScrolls[info.quest].boss.name(lang), userDamage: info.userDamage, bossDamage: info.bossDamage}, lang)}\``;
+      break;
+
+    case 'boss_dont_attack':
+      msg = `\`${shared.i18n.t('chatBossDontAttack', {username: info.user, bossName: questScrolls[info.quest].boss.name(lang), userDamage: info.userDamage}, lang)}\``;
+      break;
+
+    case 'boss_rage':
+      msg = `\`${questScrolls[info.quest].boss.rage.effect(lang)}\``;
+      break;
+
+    case 'boss_defeated':
+      msg = `\`${shared.i18n.t('chatBossDefeated', {bossName: questScrolls[info.quest].boss.name(lang)}, lang)}\``;
+      break;
+
+    case 'user_found_items':
+      foundText = _.reduce(info.items, (m, v, k) => {
+        m.push(`${v} ${questScrolls[info.quest].collect[k].text(lang)}`);
+        return m;
+      }, []).join(', ');
+      msg = `\`${shared.i18n.t('chatFindItems', {username: info.user, items: foundText}, lang)}\``;
+      break;
+
+    case 'all_items_found':
+      msg = `\`${shared.i18n.t('chatItemQuestFinish', lang)}\``;
+      break;
+
+    case 'spell_cast_party':
+      msg = `\`${shared.i18n.t('chatCastSpellParty', {username: info.user, spell: spells[info.class][info.spell].text(lang)}, lang)}\``;
+      break;
+
+    case 'spell_cast_user':
+      msg = `\`${shared.i18n.t('chatCastSpellUser', {username: info.user, spell: spells[info.class][info.spell].text(lang), target: info.target}, lang)}\``;
+      break;
+
+    case 'quest_abort':
+      msg = `\`${shared.i18n.t('chatQuestAborted', {username: info.user, questName: questScrolls[info.quest].text(lang)}, lang)}\``;
+      break;
+
+    case 'tavern_quest_completed':
+      msg = `\`${quests[info.quest].completionChat(lang)}\``;
+      break;
+
+    case 'tavern_boss_rage_tired':
+      msg = `\`${shared.i18n.t('tavernBossTired', {rageName: quests[info.quest].boss.rage.title(lang), bossName: quests[info.quest].boss.name(lang)}, lang)}\``;
+      break;
+
+    case 'tavern_boss_rage':
+      msg = `\`${quests[info.quest].boss.rage[info.scene](lang)}\``;
+      break;
+
+    case 'tavern_boss_desperation':
+      msg = `\`${quests[info.quest].boss.desperation.text(lang)}\``;
+      break;
+
+    case 'claim_task':
+      msg = `${shared.i18n.t('userIsClamingTask', {username: info.user, task: info.task}, lang)}`;
+      break;
+  }
+
+  return msg;
+}
+
+function _translateSystemMessages (group, user) {
+  let lang = user.preferences.language;
+
+  group.chat.forEach((chat, i, chatArray) => {
+    if (!_.isEmpty(chat.info)) {
+      chatArray[i].text = _translateMessage(lang, chat.info);
+    }
+  });
+
+  return group;
+}
+
+schema.statics.translateSystemMessages = _translateSystemMessages;
```

Comment:
why is this function static? Wouldn't make more senso to be a class method? so you can call group.translateSystemMessages(user) and the group will be available as `this`

Reply:
Changed to a class method here: 0360326f41485c8a9514b586dc4d9e4ef30d587c
It will break some tests, but I agree with you that it is better this way.

Now `user.preferences.language` raises an exception for me.

```
server_1  | 2019-05-15T06:37:29.319Z - error: TypeError: Cannot read property 'language' of undefined
server_1  |     at model.translateSystemMessages (/code/website/server/models/group.js:532:31)
server_1  |     at Function.groupToJSONCleanChat [as toJSONCleanChat] (/code/website/server/models/group.js:422:9)
server_1  |     at process._tickCallback (internal/process/next_tick.js:68:7)
```

To avoid the exception, for now I'm doing a verification if `user.preferences` is not null.

```javascript
let lang = user.preferences ? user.preferences.language : 'en';
```

Reply:
The issue is that for some reason the user's preferences aren't loaded from the db because they always exist so it looks likes there are cases where user.preferences aren't in the list of user fields loaded which is not fine if we need to access the property.

It's probably the spell casting route that doesn't load them